presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -492,6 +493,90 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
         }
 
+        @Override
+        public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Expression> context)
+        {
+            Expression inheritedPredicate = context.get();
+
+            // See if we can rewrite left join in terms of a plain inner join
+            if (node.getType() == SpatialJoinNode.Type.LEFT && canConvertOuterToInner(node.getRight().getOutputSymbols(), inheritedPredicate)) {
+                node = new SpatialJoinNode(node.getId(), SpatialJoinNode.Type.INNER, node.getLeft(), node.getRight(), node.getOutputSymbols(), node.getFilter());
+            }
+
+            Expression leftEffectivePredicate = effectivePredicateExtractor.extract(node.getLeft());
+            Expression rightEffectivePredicate = effectivePredicateExtractor.extract(node.getRight());
+            Expression joinPredicate = node.getFilter();
+
+            Expression leftPredicate;
+            Expression rightPredicate;
+            Expression postJoinPredicate;
+            Expression newJoinPredicate;
+
+            switch (node.getType()) {
+                case INNER:
+                    InnerJoinPushDownResult innerJoinPushDownResult = processInnerJoin(inheritedPredicate,
+                            leftEffectivePredicate,
+                            rightEffectivePredicate,
+                            joinPredicate,
+                            node.getLeft().getOutputSymbols());
+                    leftPredicate = innerJoinPushDownResult.getLeftPredicate();
+                    rightPredicate = innerJoinPushDownResult.getRightPredicate();
+                    postJoinPredicate = innerJoinPushDownResult.getPostJoinPredicate();
+                    newJoinPredicate = innerJoinPushDownResult.getJoinPredicate();
+                    break;
+                case LEFT:
+                    OuterJoinPushDownResult leftOuterJoinPushDownResult = processLimitedOuterJoin(inheritedPredicate,
+                            leftEffectivePredicate,
+                            rightEffectivePredicate,
+                            joinPredicate,
+                            node.getLeft().getOutputSymbols());
+                    leftPredicate = leftOuterJoinPushDownResult.getOuterJoinPredicate();
+                    rightPredicate = leftOuterJoinPushDownResult.getInnerJoinPredicate();
+                    postJoinPredicate = leftOuterJoinPushDownResult.getPostJoinPredicate();
+                    newJoinPredicate = leftOuterJoinPushDownResult.getJoinPredicate();
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unsupported spatial join type: " + node.getType());
+            }
+
+            verify(postJoinPredicate.equals(TRUE_LITERAL), "Spatial join cannot have post join predicate");
+
+            newJoinPredicate = simplifyExpression(newJoinPredicate);
+            verify(!newJoinPredicate.equals(BooleanLiteral.FALSE_LITERAL), "Spatial join predicate is missing");
```

Comment:
Can this happen through the normal course of expression simplification and optimization? In that case, it should turn into a regular join or even an empty ValuesNode.

Reply:
@martint Maybe... I can't think of an example though and I'm reluctant to add code without being able to test it. What do you think?

Reply:
It seems like it currently can't happen for anything other that full join due to lack of inference power. But it may start breaking later if we make the inference engine more capable.

Reply:
@martint What would you recommend? Should I try to write this code and hope it works or leave it as is?